### PR TITLE
[Merged by Bors] - feat: add Nat.digits_append_digits

### PR DESCRIPTION
**Pull request by dwrensha**
Similar to `Nat.ofDigits_digits_append_digits`, but with a `digits` on the RHS instead of an `ofDigits` on the LHS.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by dwrensha** [comment]
This is a generalization of a lemma that I needed for a USAMO problem: https://github.com/dwrensha/math-puzzles-in-lean4/blob/918bb01e426d1c60ae2f9b841dfeb95f1f9f7841/MathPuzzles/Usa2003Q1.lean#L21-L30

cc @b-mehta, who minimized the proof of `digits_add'`, and @mo271 who has also worked on the `Data.Nat.Digits` file.

File: Mathlib/Data/Nat/Digits.lean [hunk 1]
```diff
@@ -426,6 +426,19 @@ theorem ofDigits_digits_append_digits {b m n : ℕ} :
...
+    digits b n ++ digits b m = digits b (n + b ^ (digits b n).length * m) := by
+  rcases Decidable.eq_or_ne m 0 with (rfl | hm)
+  · simp
+  have : n + b ^ (digits b n).length * m = ofDigits b (digits b n ++ digits b m)
+  · simp [ofDigits_append, ofDigits_digits, add_comm, mul_comm]
+  rw [this, digits_ofDigits _ hb]
+  · simp [forall_and, or_imp]
+    constructor <;> exact (fun x hx ↦ digits_lt_base hb hx)
+  · intro hnz
+    rw[List.getLast_append' _ _ (digits_ne_nil_iff_ne_zero.mpr hm)]
```
Comment (mo271) [inline Mathlib/Data/Nat/Digits.lean:439]:
```suggestion
    rw [List.getLast_append' _ _ (digits_ne_nil_iff_ne_zero.mpr hm)]
```

**Comment by mo271** [comment]
This also works

```lean
theorem digits_append_digits {b m n : ℕ} (hb : 1 < b) :
    digits b n ++ digits b m = digits b (n + b ^ (digits b n).length * m) := by
  rw [← ofDigits_digits_append_digits]
  refine' (digits_ofDigits b hb _ (fun l hl => _) (fun h_append => _)).symm
  · rcases (List.mem_append.mp hl) with (h | h) <;> exact digits_lt_base hb h
  · by_cases digits b m = []
    · simp only [h, List.append_nil] at h_append ⊢
      exact getLast_digit_ne_zero b <| digits_ne_nil_iff_ne_zero.mp h_append
    · exact (List.getLast_append' _ _ h) ▸
          (getLast_digit_ne_zero _ <| digits_ne_nil_iff_ne_zero.mp h)
```

**Review by mo271 — APPROVED** [review]

**Comment by dwrensha** [comment]
@mo271 Thanks! Your proof looks nicer, so I switched to that.

I also weakened `hb` to only require that `0 < b`, rather than `1 < b`.
(We can't weaken it any more, because the theorem is *not* true if `b = 0`.)

**Review by mo271 — APPROVED** [review]

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**